source/service.go

```diff
@@ -221,54 +231,85 @@ func (sc *serviceSource) setResourceLabel(service v1.Service, endpoints []*endpo
 }
 
 func (sc *serviceSource) generateEndpoints(svc *v1.Service, hostname string) []*endpoint.Endpoint {
+	hostname = strings.TrimSuffix(hostname, ".")
+	ttl, err := getTTLFromAnnotations(svc.Annotations)
+	if err != nil {
+		log.Warn(err)
+	}
+
+	epA := &endpoint.Endpoint{
+		RecordTTL:  ttl,
+		RecordType: endpoint.RecordTypeA,
+		Labels:     endpoint.NewLabels(),
+		Targets:    make(endpoint.Targets, 0, defaultTargetsCapacity),
+		DNSName:    hostname,
+	}
+
+	epCNAME := &endpoint.Endpoint{
+		RecordTTL:  ttl,
+		RecordType: endpoint.RecordTypeCNAME,
+		Labels:     endpoint.NewLabels(),
+		Targets:    make(endpoint.Targets, 0, defaultTargetsCapacity),
+		DNSName:    hostname,
+	}
+
 	var endpoints []*endpoint.Endpoint
+	var targets endpoint.Targets
 
-	hostname = strings.TrimSuffix(hostname, ".")
 	switch svc.Spec.Type {
 	case v1.ServiceTypeLoadBalancer:
-		endpoints = append(endpoints, extractLoadBalancerEndpoints(svc, hostname)...)
+		targets = append(targets, extractLoadBalancerTargets(svc)...)
 	case v1.ServiceTypeClusterIP:
 		if sc.publishInternal {
-			endpoints = append(endpoints, extractServiceIps(svc, hostname)...)
+			targets = append(targets, extractServiceIps(svc)...)
 		}
 		if svc.Spec.ClusterIP == v1.ClusterIPNone {
-			endpoints = append(endpoints, sc.extractHeadlessEndpoint(svc, hostname)...)
+			endpoints = append(endpoints, sc.extractHeadlessEndpoints(svc, hostname)...)
 		}
 
 	}
+
+	for _, t := range targets {
+		if suitableType(t) == endpoint.RecordTypeA {
+			epA.Targets = append(epA.Targets, t)
+		}
+		if suitableType(t) == endpoint.RecordTypeCNAME {
+			epCNAME.Targets = append(epCNAME.Targets, t)
+		}
+	}
+
+	if len(epA.Targets) > 0 {
+		endpoints = append(endpoints, epA)
+	}
+	if len(epCNAME.Targets) > 0 {
+		endpoints = append(endpoints, epCNAME)
+	}
 	return endpoints
 }
 
-func extractServiceIps(svc *v1.Service, hostname string) []*endpoint.Endpoint {
-	ttl, err := getTTLFromAnnotations(svc.Annotations)
-	if err != nil {
-		log.Warn(err)
-	}
+func extractServiceIps(svc *v1.Service) endpoint.Targets {
```

Comment:
@dereulenspiegel Can you double-check this function. Seems to me it only returns an empty list or the cluster IP and can be simplified to:

```golang
if svc.Spec.ClusterIP != v1.ClusterIPNone {
  return endpoint.Targets{svc.Spec.ClusterIP}
}
return nil // or endpoint.Targets{}
```

Reply:
Yeah, you're right. Sometimes I don't see the woods because of all of the trees. I fixed it.

Reply:
Me too :sweat_smile: I changed it one more time to make it easier for my brain.